src/transport/PointToPointBroker.cpp

```diff
@@ -502,6 +519,87 @@ std::set<int> PointToPointBroker::getIdxsRegisteredForGroup(int groupId)
     return groupIdIdxsMap[groupId];
 }
 
+void PointToPointBroker::incrementSentMsgCount(
+  faabric::PointToPointMessage& msg,
+  int groupId,
+  int recvIdx)
+{
+    if (groupId != currentGroupId) {
+        if (currentGroupId != NO_CURRENT_GROUP_ID) {
+            SPDLOG_WARN(
+              "Changing the current group Id in PTP broker ({} -> {})",
+              currentGroupId,
+              groupId);
+        }
+        currentGroupId = groupId;
+        int groupSize = getIdxsRegisteredForGroup(groupId).size();
+        // We assume that group indexes will always go from 0 to `groupSize - 1`
+        // We also want to start counting from one. This way, an unset sequence
+        // number (0 by default in protobuf) will indicate that it is unset.
+        if (sentMsgCount != nullptr) {
+            sentMsgCount->clear();
+            sentMsgCount->resize(groupSize, 1);
+        } else {
+            sentMsgCount = std::make_unique<std::vector<int>>(groupSize, 1);
+        }
+    }
+
+    assert(recvIdx < sentMsgCount->size());
+    msg.set_seqnum(sentMsgCount->at(recvIdx)++);
+}
+
+int PointToPointBroker::getExpectedSeqNum(int groupId, int sendIdx)
+{
+    if (groupId != currentGroupId) {
+        if (currentGroupId != NO_CURRENT_GROUP_ID) {
+            SPDLOG_WARN(
+              "Changing the current group Id in PTP broker ({} -> {})",
+              currentGroupId,
+              groupId);
+        }
+        currentGroupId = groupId;
+        int groupSize = getIdxsRegisteredForGroup(groupId).size();
+        // We assume that group indexes will always go from 0 to `groupSize - 1`
+        // We also want to start counting from one. This way, an unset sequence
+        // number (0 by default in protobuf) will indicate that it is unset.
+        if (recvMsgCount != nullptr) {
+            recvMsgCount->clear();
+            recvMsgCount->resize(groupSize, 1);
+        } else {
+            recvMsgCount = std::make_unique<std::vector<int>>(groupSize, 1);
+        }
+    }
```

Comment:
Could share this with the other lazy initialisation, it would look something like:

```cpp
// Receive version (will be similar for send)
if(groupId != currentGroupId) {
    recvMsgCount = initSequenceCounter(groupId);
}
```

Reply:
I also just realised that we need to initialise both sequence counters at the same time.

In the general case when we are sending and receiving ordered messages, the `groupId != currentGroupId` check will only fail once per thread and group id. So regardless of whether we hit it sending or receiving, we need to reset all coutners.